test_plans/atomic_ref.asciidoc

```diff
@@ -505,4 +505,262 @@ Check if it returns the original value of the referenced object. Check returned
+
+* `memory_order::relaxed`
+* `memory_order::acq_rel`
+* `memory_order::seq_cst`
+
+These `SPACE` values:
+
+* `access::address_space::global_space`
+* `access::address_space::generic_space`
+
+==== Atomicity for work_group_scope
+
+This test uses a `parallel_for` with `nd_range` parameter to atomically update array, each element from all work-items in one work-group:
+
+[source,c++]
+----
+std::array<TYPE, GROUP_RANGE> vals;
+vals.fill(0);
+sycl::buffer buf{vals.data(), {GROUP_RANGE}};
+q.submit([&](sycl::handler &cgh) {
+  sycl::accessor acc{buf, cgh};
+  sycl::local_accessor<TYPE> lacc{{1}, cgh}
+  cgh.parallel_for(sycl::nd_range(GROUP_RANGE * LOCAL_SIZE, LOCAL_SIZE),
+                    [=](auto item) {
+     sycl::atomic_ref<TYPE, ORDER, SCOPE, SPACE> a_r{lacc[0]};
+     if (a_r.fetch_sub(2) == -(LOCAL_SIZE*2)) {
+        acc[item.group_id()] = a_r.load();
+     }
+  });
+});
+----
+
+Verify that `val[i] == -LOCAL_SIZE*2` after the loop completes.
+(The result must be accurate within some small error range when `TYPE` is a floating point type.)
+
+This test is repeated for all combinations of supported values listed below:
+
+These `TYPE` types:
+
+* `int`
+* `float`
+* `long long` (if `sizeof(long long) == 8` then only if the device has `aspect::atomic64`)
+* `double` (only if the device has `aspect::fp64` and if `sizeof(double) == 8` only if the device has `aspect::atomic64`)
+
+These `SCOPE` values which are supported by the device (according to `info::device::atomic_memory_scope_capabilities`):
+
+* `memory_scope::device`
+* `memory_scope::system`
+* `memory_scope::work_group`
+
+These `ORDER` values which are supported by the device (according to `info::device::atomic_memory_order_capabilities`):
+
+* `memory_order::relaxed`
+* `memory_order::acq_rel`
+* `memory_order::seq_cst`
+
+These `SPACE` values:
+
+* `access::address_space::local_space`
+* `access::address_space::generic_space`
+
+==== Aquire and release
+
+This test uses a `parallel_for` with `nd_range` parameter to atomically update shared scalars in two threads within local group.
+
+[source,c++]
+----
+const TYPE val chaged_val = 42;
+sycl::nd_range nd_range(GLOBAL_RANGE, 2);
+std::array<bool, GLOBAL_RANGE / 2> res;
+res.fill(false);
+sycl::buffer buf{res.data(), {GLOBAL_RANGE/2}};
+q.submit([&](sycl::handler &cgh) {
+  sycl::accessor res_acc{buf, cgh};
+  sycl::local_accessor<TYPE, 0> x{cgh};
+  sycl::local_accessor<TYPE, 0> y{cgh};
+  sycl::local_accessor<TYPE, 0> A{cgh};
+  sycl::local_accessor<TYPE, 0> B{cgh};
+  cgh.parallel_for(nd_range, sycl::range(2), [=](auto item) {
+    sycl::atomic_ref<TYPE, ORDER, SCOPE, SPACE> refA{A};
+    sycl::atomic_ref<TYPE, ORDER, SCOPE, SPACE> refB{B};
+    refA.store(0);
+    refB.store(0);
+    sycl::group_barrier(item.get_group());
+    if (item.get_local_id() == 0) {
+        x = refA.load();
+        refB.store(1);
+      } else {
+        y = refB.load();
+        refA.store(1);
+      }
```

Comment:
I think you need a barrier after this line:

```
sycl::group_barrier(item.get_group());
```

Otherwise, the code below could run before all work-items finish running the test.

Reply:
You are right, added.